src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureUnitTests.cs

```diff
@@ -0,0 +1,1060 @@
+    {
+        static void Main(string[] args)
+        {
+            try
+            {
+                Console.WriteLine(""Foo"");
+            }
+            catch [|(DerivedException ex)|]
+            {
+                throw;
+            }
+        }
+    }
+
+    [RequiresPreviewFeatures]
+    public class Lib
+    {
+    }
+}
+";
+
+            await new VerifyCS.Test
+            {
+                LanguageVersion = CodeAnalysis.CSharp.LanguageVersion.CSharp9,
+                TestState =
+                {
+                    Sources =
+                    {
+                        csInput
+                    }
+                },
+                MarkupOptions = MarkupOptions.UseFirstDescriptor,
+            }.RunAsync();
+        }
+
+        [Fact]
+        public async Task TestArrayOrPreviewTypes()
+        {
+            var csInput = @" 
+using System;
+namespace Preview_Feature_Scratch
+{" + setupPreviewAttribute +
+@"
+    public class Program
+    {
+        static void Main(string[] args)
+        {
+            Lib[] array = [|new Lib[] { }|];
+        }
+    }
+
+    [RequiresPreviewFeatures]
+    public class Lib
+    {
+    }
+}
+";
+
+            await new VerifyCS.Test
+            {
+                LanguageVersion = CodeAnalysis.CSharp.LanguageVersion.CSharp9,
+                TestState =
+                {
+                    Sources =
+                    {
+                        csInput
+                    }
+                },
+                MarkupOptions = MarkupOptions.UseFirstDescriptor,
+            }.RunAsync();
+        }
+
+        [Fact]
+        public async Task TestPreviewMethodBinaryOperator()
+        {
+            var csInput = @" 
+using System;
+namespace Preview_Feature_Scratch
+{" + setupPreviewAttribute +
+@"
+    public class Program
+    {
+        static void Main(string[] args)
+        {
+            var a = new Fraction();
+            var b = new Fraction();
+            b = [|b + a|];
+        }
+    }
+
+    public readonly struct Fraction
+    {
+        [RequiresPreviewFeatures]
+        public static Fraction operator +(Fraction a, Fraction b) => a;
+    }
+}
+";
+
+            await new VerifyCS.Test
+            {
+                LanguageVersion = CodeAnalysis.CSharp.LanguageVersion.CSharp9,
+                TestState =
+                {
+                    Sources =
+                    {
+                        csInput
+                    }
+                },
+                MarkupOptions = MarkupOptions.UseFirstDescriptor,
+            }.RunAsync();
+        }
+
+        [Fact]
+        public async Task TestPreviewMethodCallingPreviewMethod()
+        {
+            var csInput = @" 
+using System;
+namespace Preview_Feature_Scratch
+{" + setupPreviewAttribute +
+@"
+public class Program
+{
+    [RequiresPreviewFeatures]
+    public virtual void PreviewMethod()  { }
+   
+    [RequiresPreviewFeatures]
+    void CallSite()
+    {
+        PreviewMethod();
```

Comment:
This test will fail right now because I don't know what the right way to detect this is. If I inspect `IOperation` here, it will of type `IInvocation`. My plan was to go up the parent chain here to inspect the `CallSite` method declaration and check if the `reqeuiresPreviewFeatures` attribute had been set. However I'm running into problems. The parent chain here (starting from `PreviewMethod()`) seems to be `ExpressionOperation -> BlockOperation -> MethodBodyOperation`. `MethodBodyOperation`'s parent in `null`, so I'm unable to get to the `CallSite`'s definition. However, if I inspect the `Syntax` property of `MethodBodyDeclaration` and traverse it's parent, I can get to `CallSite`'s syntax node. This would be C# specific however, so it is not an option. Is there another API that let's me inspect `CallSite` properly?

@mavasani @buyaa-n

Reply:
Just calling this comment out again. I'm wondering if you guys know the answer or a better way here? @mavasani @buyaa-n

Reply:
> My plan was to go up the parent chain here to inspect the CallSite method declaration and check if the reqeuiresPreviewFeatures attribute had been set. However I'm running into problems. The parent chain here (starting from PreviewMethod()) seems to be ExpressionOperation -> BlockOperation -> MethodBodyOperation. MethodBodyOperation's parent in null, so I'm unable to get to the CallSite's definition.

You should check the ContainingSymbol, not the parent, see how it is done in PCA:
https://github.com/dotnet/roslyn-analyzers/blob/582230ea924cba9f2e2240b852b12048c07aa31c/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.cs#L1050-L1070 
Here on row 1050 checks if the operation has platform attribute, if it has then loading the containing symbol's (i. e callsite) attributes on row 1058, then on row 1061 checking if the call site attributes can suppress the operation attributes

Note that here the `TryGetOrCreatePlatformAttributes` has a recursive call and collects all attributes from symbol to assembly, not only go up one level

Reply:
> You should check the ContainingSymbol, not the parent, see how it is done in PCA

This is what the code is doing now, but it doesn't work I think. For this test case:

``` csharp
public class Program
{
    [RequiresPreviewFeatures]
    public virtual void PreviewMethod()  { }
   
    [RequiresPreviewFeatures]
    void CallSite()
    {
        PreviewMethod();
    }
}
}
```

`PreviewMethod()` (from inside CallSite) will be accessed as an `INamedTypeSymbol`. It's `ContainingType/Symbol` will be `Program`, not `CallSite`, because `Program` is the one that defines `PreviewMethod`. What I need is a way to know that the current `PreviewMethod` invocation has been called from within another method that is also marked as Preview. I wonder if I should be using SemanticModel here. I'll explore locally

Reply:
> PreviewMethod() (from inside CallSite) will be accessed as an INamedTypeSymbol. It's ContainingType/Symbol will be Program, not CallSite, because Program is the one that defines PreviewMethod.

That doesn't sound right, the `ContainingType` of the PreviewMethod() invocation is the `CallSite()` method, not `Program`, `Program` is the  `ContainingType` of the `PreviewMethod` declaration

Reply:
`PreviewMethod()` invocation should be flagged from operation analysis not from symbol analysis, probably there is duplication in your analysis, anyway, I believe the diagnostic is being reported in the `PreviewMethod()` invocation not in the declaration, so from there you should be able to get the real `CallSite()` using `ContainingType`

Reply:
Calling context.ContainingSymbol worked. This is now resolved.